File: 프로그래머스/lv3/64064. 불량 사용자/불량 사용자.py
from itertools import permutations

def check(user, ban):
    if len(user) != len(ban): # 길이 다르면 비교X
        return False
    else:
        for i, j in zip(user, ban):
            if j == '*':
                continue
            if i != j:
                return False
        return True

def solution(user_id, banned_id):
    answer = []
    
    for i in permutations(user_id, len(banned_id)):
        count = 0
        for a, b in zip(i, banned_id):
            # print(a, b)
            if check(a, b):
                count += 1
        
        if count == len(banned_id):
            if set(i) not in answer:
                answer.append(set(i))
    
    return len(answer)

File: 프로그래머스/lv3/64064. 불량 사용자/불량 사용자.swift
import Foundation

// 비교 메서드
func compare(_ target: String, _ banned: String) -> Bool {
    if target.count != banned.count { return false }
    
    for (i, j) in zip(target, banned) {
        if i == j || j == "*" { continue }
        if i != j { return false }
    }
    
    return true
}

// 순열
func permutation<T: Comparable>(_ target: [T], _ n: Int) -> [[T]] {
    var ret = [[T]]()
    var visited = [Bool](repeating: false, count: target.count)
    
    func recur(_ cur: [T]) {
        if cur.count == n { ret.append(cur); return }
        
        for i in 0..<target.count {
            if visited[i] { continue }
            
            visited[i] = true
            recur(cur + [target[i]])
            visited[i] = false
        }
    }
    recur([])
    return ret
}

func solution(_ user_id:[String], _ banned_id:[String]) -> Int {
    var answer = Set<Set<String>>()
    
    for users in permutation(user_id, banned_id.count) {
        var count = 0
        for (a, b) in zip(users, banned_id) {
            if compare(a, b) {
                count += 1
            }
        }
        
        if count == banned_id.count {
            answer.insert(Set(users))
        }
    }
    
    return answer.count
}